net/tstun/mtu.go

```diff
@@ -1,33 +1,145 @@
+// headers above the link layer (e.g. IP headers, UDP headers, Wireguard
+// headers, etc.). We have to think about several different values of MTU:
+//
+// Wire MTU: The MTU of an interface underneath the tailscale TUN, e.g. an
+// Ethernet network card will default to a 1500 byte MTU. The user may change
+// this MTU at any time.
+//
+// TUN MTU: The current MTU of the tailscale TUN. This MTU is adjusted downward
+// to make room for the wireguard/tailscale headers. For example, if the
+// underlying network interface's MTU is 1500 bytes, the maximum size of a
+// packet entering the tailscale TUN is 1420 bytes. The user may change this MTU
+// at any time via the OS's tools (ifconfig, ip, etc.).
+//
+// User configured initial MTU: The MTU the tailscale TUN should be created
+// with, set by the user via TS_DEBUG_MTU. It should be adjusted down from the
+// underlying interface MTU by 80 bytes to make room for the wireguard
+// headers. This envknob is mostly for debugging. This value is used once at TUN
+// creation and ignored thereafter.
+//
+// User configured current MTU: The MTU set via the OS's tools (ifconfig, ip,
+// etc.). This MTU can change at any time. Setting the MTU this way goes through
+// the MTU() method of tailscale's TUN wrapper.
+//
+// Maximum probed MTU: This is the largest MTU size that we send probe packets
+// for.
+//
+// Safe MTU: If the tailscale TUN MTU is set to this value, almost all packets
+// will get to their destination. Tailscale defaults to this MTU in the absence
+// of path MTU probe information or user MTU configuration. We may occasionally
+// find a path that needs a smaller MTU but it is very rare.
+//
+// Peer MTU: This is the path MTU to a peer's current best endpoint. It defaults
+// to the Safe MTU unless we have path MTU probe results that tell us otherwise.
+//
+// Initial MTU: This is the MTU tailscaled creates the TUN with. In order of
+// priority, it is:
+//
+// 1. If set, the value of TS_DEBUG_MTU clamped to a maximum of 65536
+// 2. If TS_DEBUG_ENABLE_PMTUD is set, the maximum size MTU we probe, minus wg
+//    overhead
+// 3. If TS_DEBUG_ENABLE_PMTUD is not set, the Safe MTU
+//
+// Current MTU: This the MTU of the tailscale TUN at any given moment
+// after TUN creation. In order of priority, it is:
+//
+// 1. The MTU set by the user via the OS, if it has ever been set
+// 2. If TS_DEBUG_ENABLE_PMTUD is set, the maximum size MTU we probe, minus wg
+//    overhead
+// 4. If TS_DEBUG_ENABLE_PMTUD is not set, the Safe MTU
+
+// TunMTU is the MTU for the tailscale TUN.
+type TunMTU uint32
+
+// WireMTU is the MTU for the underlying network devices.
+type WireMTU uint32
 
 const (
-	maxMTU     uint32 = 65536
-	defaultMTU uint32 = 1280
+	// maxTunMTU is the largest MTU we will consider for the Tailscale
+	// TUN. This is inherited from wireguard-go and can be surprisingly
+	// small; on Windows it is currently 2048 - 32 bytes.
```

Comment:
We don't currently use the rio Bind on Windows, we use our own conn.Bind in https://github.com/tailscale/tailscale/blob/main/wgengine/magicsock/magicsock.go#L2018

Reply:
That said, the value below bottoms out in: 

```
device/queueconstants_android.go:       MaxSegmentSize             = 2200
device/queueconstants_default.go:       MaxSegmentSize             = (1 << 16) - 1 // largest possible UDP datagram
device/queueconstants_ios.go:const MaxSegmentSize = 1700
device/queueconstants_windows.go:       MaxSegmentSize             = 2048 - 32 // largest possible UDP datagram
```

So iOS is also rather low and I do not know/remember where the 1700 value came from. For one thing we need to remain somewhat conservative with allocation on iOS and we have a desire to adjust the buffer pooling strategy so we can eventually increase these, but without substantially increasing our active pool watermark. The last time we were thinking about this substantially, we were considering a dependency on gvisors buffer infrastructure (which is skbuff like in capabilities), though there are necessary discussions about that and it's relationship to pooling on the wireguard-go side.

Reply:
Yeah, Windows uses it in the rio Bind that doesn't affect us, but the value also propagated elsewhere in Windows code that does affect us. I think that plan sounds good.